gpkit/signomial_program.py

```diff
@@ -1,179 +1,184 @@
-
-        # add subbed-out monomials at the end
-        if not exps[-1]:
-            self.A.append(0, len(self.exps)-1, 0)
-        self.A.update_shape()
-
-        self.missingbounds = missingbounds
-        if printing:
-            self.checkbounds()
-
-        return cs, exps, varlocs, self.A, p_idxs, k, unsubbedexps, unsubbedvarlocs
-
-    def localsolve(self, solver=None, printing=True, x0={}, reltol=1e-4, *args, **kwargs):
-        self.reltol = 1e-4
-        if printing:
-            print("Beginning signomial solve.")
 
-        self.x0 = x0
-        self.sp_iters = 0
-        self.presolve = self.last
-        self.subbedcost = self.cost.sub(self.substitutions)
-        sol = self._solve(solver, printing=printing, *args, **kwargs)
-        if printing:
-            print("Solving took %i GP solves." % self.sp_iters)
-        return sol
-
-    def solve(self, *args, **kwargs):
-        raise KeyError("Signomial programs have only local solutions, and"
-                       "must be solved using the the 'localsolve' function"
-                       "instead of the 'solve' function.")
+from .geometric_program import GeometricProgram
+
+
+class SignomialProgram(object):
+    """Prepares a collection of signomials for a SP solve.
+
+    Arguments
+    ---------
+    cost : Constraint
+        Signomial to minimize when solving
+    constraints : list of Posynomials
+        Constraints to maintain when solving (implicitly Signomials <= 1)
+
+    Examples
+    --------
+    >>> gp = gpkit.geometric_program.SignomialProgram(
+                        # minimize
+                        x,
+                        [   # subject to
+                            1/x - y/x,  # <= 1, implicitly
+                            y/10  # <= 1
+                        ])
+    >>> gp.solve()
+    """
+
+    def __init__(self, cost, constraints):
+        self.cost = cost
+        self.constraints = constraints
+        self.signomials = [cost] + list(constraints)
+
+        self.posynomials, self.negynomials = [], []
+        self.negvarkeys = set()
+        for sig in self.signomials:
+            p_exps, p_cs = [], []
+            n_exps, n_cs = [{}], [1]
+            for c, exp in zip(sig.cs, sig.exps):
+                if c > 0:
+                    p_cs.append(c)
+                    p_exps.append(exp)
+                elif c < 0:
+                    n_cs.append(-c)
+                    n_exps.append(exp)
+                    self.negvarkeys.update(exp.keys())
+            posy = Posynomial(p_exps, p_cs) if p_cs != [] else None
+            negy = Posynomial(n_exps, n_cs) if n_cs != [1] else None
+            self.posynomials.append(posy)
+            self.negynomials.append(negy)
+
+        if not self.negvarkeys:
+            raise ValueError("SignomialPrograms must contain coefficients"
+                             " less than zero.")
+
+    def solve(self, solver=None, verbosity=1, x0={}, reltol=1e-4,
```

Comment:
I really liked that the difference required a user to know the mathematical properties of their model! I'd vote for keeping the "specify solve/localsolve" correctly requirement, and implementing at the `Model` level.

But willing to back down if that doesn't match your vision.